command/hook_pending.go

```diff
@@ -0,0 +1,121 @@
+package command
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform/terraform"
+	"github.com/mitchellh/cli"
+	"github.com/mitchellh/colorstring"
+)
+
+type pendingEvent struct {
+	// Kind of operation that started, or "" if the operation ended.
+	op string
+	// Human ID of the node being operated on.
+	id string
+}
+
+type PendingHook struct {
+	terraform.NilHook
+
+	Colorize *colorstring.Colorize
+	Ui       cli.Ui
+
+	events chan *pendingEvent
+}
+
+func (h *PendingHook) PreApply(
+	n *terraform.InstanceInfo,
+	s *terraform.InstanceState,
+	d *terraform.InstanceDiff) (terraform.HookAction, error) {
+	op := "modifying"
+	if d.Destroy {
+		op = "destroying"
+	} else if s.ID == "" {
+		op = "creating"
+	}
+
+	h.events <- &pendingEvent{
+		op: op,
+		id: n.HumanId(),
+	}
+
+	return terraform.HookActionContinue, nil
+}
+
+func (h *PendingHook) PostApply(
+	n *terraform.InstanceInfo,
+	s *terraform.InstanceState,
+	applyerr error) (terraform.HookAction, error) {
+	h.events <- &pendingEvent{
+		id: n.HumanId(),
+	}
+
+	return terraform.HookActionContinue, nil
+}
+
+func (h *PendingHook) ShowPendingOperationsInBackground(doneCh <-chan struct{}) {
+	h.events = make(chan *pendingEvent)
+	idToOp := make(map[string]string)
+
+	go func() {
+		for {
+			select {
+			case <-doneCh:
+				// The apply is done; nothing more to print.
+				return
+
+			case event := <-h.events:
+				// Something happened! Update our internal state and restart the timer.
+				if event.op == "" {
+					// PostApply. Note that this can get called even if there's a no-op diff.
+					delete(idToOp, event.id)
+				} else {
+					// PreApply.
+					if _, found := idToOp[event.id]; found {
+						panic(fmt.Sprintf("got duplicate PreApply for %s", event.id))
```

Comment:
Turns out this panic can fire. Working on an updated version which refcounts calls to Pre/PostApply.